### PR TITLE
Save changes before closing interaction.

### PR DESCRIPTION
**Pull request by emilhammarstedtst**
Screenshot from 2023-02-09 13-44-36

#### What it does
- Question to save all changes, cancel or discard all changes before closing
- Display what is not yet saved.
- Reorder options and reword the question and options to be clear of the outcome of each action
- Implemented using ConfirmDialogSave extending ConfirmDialog

#### How to test

Electron build of theia and make one or more editors dirty then click the "X" to close the theia window. the popup should appear with a question to save.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Contributed by STMicroelectronics

### CODE REVIEW

File: packages/core/src/browser/dialogs.ts [hunk 1]
```diff
@@ -401,6 +401,68 @@ export async function confirmExit(): Promise<boolean> {
...
+    }
+
+}
+
+export async function confirmExitWithOrWithoutSaving(captionsToSave: string[], performSave: () => void): Promise<boolean> {
+    let msg: string | HTMLElement = nls.localize('theia/core/quitMessage', 'Any unsaved changes will be lost.');
+
+    if (typeof captionsToSave !== 'undefined') {
+        const div = document.createElement('div');
+        div.innerText = nls.localize('theia/core/quitMessageFileList', 'The following have not yet been saved') + ':';
```
Comment (vince-fugnitto) [inline packages/core/src/browser/dialogs.ts:441]: I believe we can remove this `innerText`.

File: packages/core/src/browser/dialogs.ts [hunk 2]
```diff
@@ -401,6 +401,68 @@ export async function confirmExit(): Promise<boolean> {
...
+            span.appendChild(b);
+            span.appendChild(document.createElement('br'));
+        });
+        span.appendChild(document.createElement('br'));
+        div.appendChild(span);
+        msg = div;
+    }
+
+    const safeToExit = await new ConfirmDialogSave({
+        title: nls.localizeByDefault('Are you sure you want to quit?'),
```
Comment (vince-fugnitto) [inline packages/core/src/browser/dialogs.ts:457]: A default localization exists and we can pass the count:

```suggestion
        title: nls.localizeByDefault('Do you want to save the changes to the following {0} files?', count),
```

File: packages/core/src/browser/dialogs.ts [hunk 3]
```diff
@@ -401,6 +401,68 @@ export async function confirmExit(): Promise<boolean> {
...
+        msg: msg,
+        ok: nls.localize('theia/core/quitOptionWithSaving', 'Quit without saving'),
```
Comment (vince-fugnitto) [inline packages/core/src/browser/dialogs.ts:459]:
```suggestion
        ok: nls.localizeByDefault("Don't Save')
```

File: packages/core/src/browser/dialogs.ts [hunk 4]
```diff
@@ -401,6 +401,68 @@ export async function confirmExit(): Promise<boolean> {
...
+        save: nls.localize('theia/core/quitOptionWithoutSaving', 'Save all and quit'),
```
Comment (vince-fugnitto) [inline packages/core/src/browser/dialogs.ts:460]:
```suggestion
        save: nls.localizeByDefault('Save All'),
```

File: packages/core/src/browser/dialogs.ts [hunk 5]
```diff
@@ -401,6 +401,68 @@ export async function confirmExit(): Promise<boolean> {
...
+        this.saveButton.classList.add('main');
+        return this.saveButton;
+    }
+
+}
+
+export async function confirmExitWithOrWithoutSaving(captionsToSave: string[], performSave: () => void): Promise<boolean> {
+    let msg: string | HTMLElement = nls.localize('theia/core/quitMessage', 'Any unsaved changes will be lost.');
+
+    if (typeof captionsToSave !== 'undefined') {
```
Comment (vince-fugnitto) [inline packages/core/src/browser/dialogs.ts:439]: In which case will `captionsToSave` be `undefined`?

File: packages/core/src/browser/dialogs.ts [hunk 6]
```diff
@@ -401,6 +401,68 @@ export async function confirmExit(): Promise<boolean> {
...
+        this.saveButton = this.createButton(text);
+        this.controlPanel.appendChild(this.saveButton);
+        this.saveButton.classList.add('main');
+        return this.saveButton;
+    }
+
+}
+
+export async function confirmExitWithOrWithoutSaving(captionsToSave: string[], performSave: () => void): Promise<boolean> {
+    let msg: string | HTMLElement = nls.localize('theia/core/quitMessage', 'Any unsaved changes will be lost.');
```
Comment (vince-fugnitto) [inline packages/core/src/browser/dialogs.ts:437]: There is a default localization we can use:

```suggestion
    let msg: string | HTMLElement = nls.localizeByDefault("Your changes will be lost if you don't save them.");
```

Comment (vince-fugnitto) [inline packages/core/src/browser/dialogs.ts:437]: I'm not sure I follow the assignment to `| HTMLElement`.

File: packages/core/src/browser/dialogs.ts [hunk 7]
```diff
@@ -401,6 +401,68 @@ export async function confirmExit(): Promise<boolean> {
...
+        super(props);
+        this.contentNode.appendChild(this.createMessageNode(this.props.msg));
+        // reorder buttons
+        this.controlPanel.childNodes.forEach(child => this.controlPanel.removeChild(child));
+        [this.acceptButton, this.closeButton].forEach(child => {
+            if (typeof child !== 'undefined') {
+                this.controlPanel.appendChild(child);
+            }
+        });
+        this.appendSaveButton(props.save).addEventListener('click', () => { props.performSave(); this.acceptButton?.click(); });
```
Comment (vince-fugnitto) [inline packages/core/src/browser/dialogs.ts:424]: We should break up the callback for clarity:

```suggestion
        this.appendSaveButton(props.save).addEventListener('click', () => { 
        props.performSave(); 
        this.acceptButton?.click(); 
     });
```

**Review by vince-fugnitto — COMMENTED** [review]
@emilhammarstedtst thank you for the contribution :+1: 

I have some general comments about the dialog, mainly to align with the one present in VS Code and re-use some already existing localizations. 

I believe that it is incorrect to default to the "Quit Without Saving" button which is selected by default. Pressing <kbd>Enter</kbd> should probably `Save All` instead similarly to VS Code.

---

Please also be sure to properly fill out the pull-request description with "How To Test", for example I don't believe these changes apply to the browser as you have them today?

File: packages/core/src/browser/dialogs.ts (hunk 6 above)
Comment (emilhammarstedtst) [reply packages/core/src/browser/dialogs.ts:437]: Sorry old leftover code. it is meant as a fallback strategy if captionsToSave was left undefined. I'll clean it up a bit.

**Comment by emilhammarstedtst** [comment]
> @emilhammarstedtst thank you for the contribution +1
> 
> I have some general comments about the dialog, mainly to align with the one present in VS Code and re-use some already existing localizations.
> 
> I believe that it is incorrect to default to the "Quit Without Saving" button which is selected by default. Pressing Enter should probably `Save All` instead similarly to VS Code.
> 
> Please also be sure to properly fill out the pull-request description with "How To Test", for example I don't believe these changes apply to the browser as you have them today?

Amended the commit with your suggestions and cleared up the "undefined"-leftover.
However the title now potentially has the the text " following 0 files?",  so i created a fallback case for this.

I dont believe the browser is in question as it is bound to use the onbeforeunload method?

**Review by msujew — APPROVED** [review]
Looks good to me 👍 I have 2 minor comments before merging though.

**Comment by vince-fugnitto** [comment]
@msujew do we want to merge for today's release?

**Pull request merged**